core/os/flock/flock.go

```diff
@@ -18,112 +18,134 @@ import (
+func Lock(n string) (*Mutex, error) {
+	m, err := New(n)
 	if err != nil {
 		return nil, err
 	}
-	l.Lock()
-	return l, nil
+	m.Lock()
+	return m, nil
 }
 
-// TryLockDevice creates a DeviceFLock and use it to acquires the lock for the
-// device specified by the given serial or name of the device in a Non-Blocking
-// way, returns the created DeviceFLock, boolean to show whether the lock has
-// been acquired or not, and error.
-func TryLockDevice(sn string) (*DeviceFLock, bool, error) {
-	l, err := NewDeviceFLock(sn)
+// TryLock creates a Mutex with the given string used as the name of the underlying
+// file, and use the Mutex to acqure the lock backed by that underlying file in
+// a non-blocking way. Returns the created Mutex and error if there are any.
+func TryLock(n string) (*Mutex, error) {
+	m, err := New(n)
 	if err != nil {
-		return nil, false, err
+		return nil, err
 	}
-	locked := l.TryLock()
-	return l, locked, nil
+	m.TryLock()
+	return m, nil
 }
 
-// DeviceFLock is a file-based lock for devices valid for a fixed peroid of
-// time, specificlly for locking the use of devices. I.e, one device can have
-// only one FLock, and after a fixed period of time, the lock will be invalid,
-// so others can acquire the FLock to use the corresponding device.
-type DeviceFLock struct {
-	sn string
-	fl *fLock
-	m  sync.Mutex
+// Mutex is a file based mutex. Mutex should only be created by calling New.
+type Mutex struct {
+	filename string
+	fl       *fLock
+	m        sync.Mutex
 }
 
-// NewDeviceFLock creates a new File-lock based lock for the device
-// identified by the given serial/name.
-func NewDeviceFLock(sn string) (*DeviceFLock, error) {
-	l := &DeviceFLock{
-		sn: sn,
-		fl: &fLock{},
+// New creates a new file based Mutex which uses the given string as the name
+// of the underlying file. The newly created Mutex is to be initiated in
+// 'unlocked' state.
+func New(n string) (*Mutex, error) {
+	m := &Mutex{
+		filename: n,
+		fl:       &fLock{},
 	}
 	var err error
-	l.fl.path, err = deviceFLockPath(sn)
-	return l, err
+	m.fl.path, err = lockFilePath(n)
+	return m, err
+}
+
+// Locked returns whether the Mutex holds the lock.
+func (m *Mutex) Locked() bool {
+	return m.fl.locked
 }
 
 // TryLock acquires the lock in a non-blocking way, returns true if the lock is
-// acquired by the DeviceFLock, false otherwise.
-func (l *DeviceFLock) TryLock() bool {
-	l.m.Lock()
-	defer l.m.Unlock()
-	locked, err := l.fl.tryLock()
+// successfully acquired by the Mutex, false otherwise. If the Mutex already
+// holds the lock, returns true.
+func (m *Mutex) TryLock() bool {
+	m.m.Lock()
+	defer m.m.Unlock()
+	err := m.fl.tryLock()
 	if err != nil {
 		return false
 	}
-	return locked
+	return true
 }
 
-// Lock acquires the lock in a blocking way.
-func (l *DeviceFLock) Lock() {
-	l.m.Lock()
-	defer l.m.Unlock()
-	l.fl.lock()
+// Lock acquires the lock in a blocking way. Locking an already locked Mutex
+// returns immediately.
```

Comment:
If locking an already locked mutex just instantly returns, aren't unlocks likely to unlock some other processes lock? Shouldn't we have a lock count?

Reply:
I think I should change it to 'Locking a Mutex which already held the lock returns immediately'.

Or do you think I should make is a recursive lock?